Fix NULL conversion warnings found in Alpine x64 Source-Build

jkoritzinsky opened the pull request:
Fixes https://github.com/dotnet/source-build/issues/4298

jkoritzinsky committed:
Fix NULL conversion warnings found in Alpine x64 Source-Build

mangod9 commented on src/coreclr/binder/customassemblybinder.cpp:
```diff
@@ -242,21 +242,21 @@ void CustomAssemblyBinder::PrepareForLoadContextRelease(INT_PTR ptrManagedStrong
 {
     m_pDefaultBinder = NULL;
-    m_ptrManagedStrongAssemblyLoadContext = NULL;
+    m_ptrManagedStrongAssemblyLoadContext = 0;
```
I worry that this kind of change reduces code readability? Are these new warnings related to a tooling update?

jkoritzinsky replied:
With a more recent musl libc (the version in Alpine 3.19), NULL is defined as nullptr directly, making what was suppressable warnings with our current toolsets into unsupressable compiler errors. This PR includes fixes for all the locations that were now procuding errors for an x64 build on Alpine 3.19 (what the source-build pipeline tests).

In the future, I plan to follow up with re-enabling the warnings and fixing the remaining cases

mangod9 replied:
so is there no way to #ifdef NULL to 0 somehow for this platform? 

cc: @AaronRobinsonMSFT @janvorli

janvorli replied:
This PR uses `(UINT_PTR)NULL` instead of 0 in a similar case, it seems like it would be better here too.

janvorli replied:
Here it would be actually `(INT_PTR)NULL`

jkoritzinsky commented:
Okay this PR has now included all of the various different changes to handle the three different definitions of `NULL` that the standard libraries we build against use (and how they interact with our DAC-ized pointer types).

- `#define NULL nullptr` as used by newer versions of musl libc (Alpine 3.19)
- `#define NULL __null` as used by glibc (GCC/Clang)
- `#define NULL 0` as used by MSVC

This generally involved changing code like `TADDR foo = NULL` to `TADDR foo = (TADDR)NULL` or `TADDR foo = 0` and changing `PTR_Bar foo = NULL` to `PTR_Bar foo = nullptr`.

jkoritzinsky committed:
Use delgating constructors to simplify DAC pointer types.

jkoritzinsky committed:
Consolidate code in the DACized pointer types and fix the nullptr/NULL problem with them

jkoritzinsky committed:
Define nullptr helpers for our checked objref types like our dacized ptr types

jkoritzinsky committed:
Fixes for the glibc-based build

jkoritzinsky committed:
Convert some of the 0s back to NULLs

jkoritzinsky commented:
I've taken a different approach here that should still work and cut the diff in half (and removed the majority of the changes required for non Alpine).

janvorli reviewed (approved):
LGTM, thank you!

AaronRobinsonMSFT commented on src/coreclr/debug/daccess/daccess.cpp:
```diff
@@ -4324,7 +4324,7 @@ ClrDataAccess::TranslateExceptionRecordToNotification(
     ULONG32 notifyType = 0;
     DWORD catcherNativeOffset = 0;
-    TADDR nativeCodeLocation = NULL;
+    TADDR nativeCodeLocation = (TADDR)NULL;
```
Since you have found all of these... thoughts on doing a `TADDR::Null` static? I am making this suggestion based on the assumption that we going to stuck with this `NULL` casting for `TADDR` for a long time.

jkoritzinsky replied:
`TADDR` is a typedef from `uintptr_t` so we can't do anything like that today.

In the future, I'd like to move in that direction (by making `TADDR` an enum-class or something like that to get us more explicit typing without regressions on Windows) but that's going to be a lot more work than I want to do in this PR.

AaronRobinsonMSFT replied:
Oh, right. I was thinking it was a `struct`. All good.

AaronRobinsonMSFT commented on src/coreclr/vm/clrex.h:
```diff
@@ -934,7 +934,7 @@ LONG CLRNoCatchHandler(EXCEPTION_POINTERS* pExceptionInfo, PVOID pv);
 
 inline CLRException::CLRException()
-  : m_throwableHandle(NULL)
+  : m_throwableHandle((OBJECTHANDLE)NULL)
```
```suggestion
  : m_throwableHandle{}
```

AaronRobinsonMSFT commented on src/coreclr/vm/codeversion.cpp:
```diff
@@ -60,7 +60,7 @@ NativeCodeVersionNode::NativeCodeVersionNode(
     unsigned ilOffset)
     :
-    m_pNativeCode(NULL),
+    m_pNativeCode((PCODE)NULL),
```
```suggestion
    m_pNativeCode{},
```

AaronRobinsonMSFT commented on src/coreclr/vm/custommarshalerinfo.cpp:
```diff
@@ -25,7 +25,7 @@ CustomMarshalerInfo::CustomMarshalerInfo(LoaderAllocator *pLoaderAllocator, Type
 , m_hndManagedType(hndManagedType)
 , m_pLoaderAllocator(pLoaderAllocator)
-, m_hndCustomMarshaler(NULL)
+, m_hndCustomMarshaler(0)
```
```suggestion
, m_hndCustomMarshaler{}
```

AaronRobinsonMSFT commented on src/coreclr/vm/domainassembly.cpp:
```diff
@@ -37,8 +37,8 @@ DomainAssembly::DomainAssembly(PEAssembly* pPEAssembly, LoaderAllocator* pLoader
     m_loading(TRUE),
-    m_hExposedModuleObject(NULL),
-    m_hExposedAssemblyObject(NULL),
+    m_hExposedModuleObject((LOADERHANDLE)NULL),
+    m_hExposedAssemblyObject((LOADERHANDLE)NULL),
```
```suggestion
    m_hExposedModuleObject{},
    m_hExposedAssemblyObject{},
```

AaronRobinsonMSFT commented on src/coreclr/vm/exceptionhandling.h:
```diff
@@ -132,7 +132,7 @@ struct ExceptionTrackerBase
     ExceptionTrackerBase(PTR_EXCEPTION_RECORD pExceptionRecord, PTR_CONTEXT pExceptionContext, PTR_ExceptionTrackerBase pPrevNestedInfo) :
         m_pPrevNestedInfo(pPrevNestedInfo),
-        m_hThrowable(NULL),
+        m_hThrowable(0),
```
```suggestion
        m_hThrowable{},
```

AaronRobinsonMSFT commented on src/coreclr/vm/exceptionhandling.h:
```diff
@@ -265,7 +265,7 @@ class ExceptionTracker : public ExceptionTrackerBase
         ExceptionTrackerBase(pExceptionRecord, pContextRecord, PTR_NULL),
         m_pThread(GetThread()),
-        m_uCatchToCallPC(NULL),
+        m_uCatchToCallPC(0),
```
```suggestion
        m_uCatchToCallPC{},
```

AaronRobinsonMSFT commented on src/coreclr/vm/exinfo.cpp:
```diff
@@ -317,7 +317,7 @@ ExInfo::ExInfo(Thread *pThread, EXCEPTION_RECORD *pExceptionRecord, CONTEXT *pEx
     m_passNumber(1),
     m_idxCurClause(0xffffffff),
-    m_notifyDebuggerSP(NULL),
+    m_notifyDebuggerSP(0),
```
```suggestion
    m_notifyDebuggerSP{},
```

AaronRobinsonMSFT commented on src/coreclr/vm/listlock.h:
```diff
@@ -65,7 +65,7 @@ class ListLockEntryBase
         m_dwRefCount(1),
         m_hrResultCode(S_FALSE),
-        m_hInitException(NULL),
+        m_hInitException((LOADERHANDLE)0),
```
```suggestion
        m_hInitException{},
```

AaronRobinsonMSFT commented on src/coreclr/vm/multicorejit.h:
```diff
@@ -92,7 +92,7 @@ class MulticoreJitCodeInfo
 
 public:
-    MulticoreJitCodeInfo() : m_entryPointAndTierInfo(NULL)
+    MulticoreJitCodeInfo() : m_entryPointAndTierInfo(0)
```
```suggestion
    MulticoreJitCodeInfo() : m_entryPointAndTierInfo{}
```

AaronRobinsonMSFT reviewed (approved):
One suggestion and a few nits to consider. Thanks!

jkoritzinsky committed:
Apply suggestions

jkoritzinsky committed:
MSVC requires explicitly-defaulted default constructors when using delegating constructors.

Also remove the explicit bool cast as it breaks comparisons against NULL with MSVC's definition.

jkoritzinsky committed:
Fix typo in comment about NULL pattern

jkoritzinsky committed:
Fix cast
